netcon/onlinedirectipclient/odtclient.cpp

```diff
@@ -0,0 +1,1822 @@
+            memset(chan_info[i].name, 0, 19);
+            strcpy(chan_info[i].origname, tokp);
+            strcpy(chan_info[i].name, tokp + 1); // skip the #
+            for (int l = 0; l < 18; l++) {
+              if (chan_info[i].name[l] == '_')
+                chan_info[i].name[l] = ' ';
+              if (chan_info[i].name[l] == '\0')
+                chan_info[i].name[l] = ' ';
+            }
+            pcount++;
+            char *ptopic = pcount;
+            while (isdigit(*ptopic)) {
+              ptopic++;
+            }
+            *ptopic = '\0';
+            ptopic++;
+            // count
+            strcpy(chan_info[i].count, pcount);
+            // topic
+            strcpy(chan_info[i].topic, ptopic);
+            // Now we want it formatted nicely.....
+            char chan_name[100];
+            memset(chan_name, ' ', 99);
+            chan_name[99] = '\0';
+            char count_list[10];
+            memset(count_list, ' ', 9);
+            count_list[9] = '\0';
+
+            memset(fmtchan, 0, 500);
+
+            int charpos = 0;
+            int endpos;
+#define LIST_ROW1 90
+#define LIST_ROW2 130
+
+            strcpy(chan_name, chan_info[i].name);
+            endpos = strlen(chan_info[i].name);
+            chan_name[endpos] = ' ';
+
+            strcpy(count_list, chan_info[i].count);
+            endpos = strlen(chan_info[i].count);
+            count_list[endpos] = ' ';
+
+            while (DLLgrtext_GetTextLineWidth(fmtchan) < LIST_ROW1) {
+              endpos = strlen(fmtchan);
+              fmtchan[endpos] = chan_name[charpos];
+              fmtchan[endpos + 1] = NULL;
+              charpos++;
+            }
+            strcat(fmtchan, "\t\0");
+            charpos = 0;
+            while (DLLgrtext_GetTextLineWidth(fmtchan) < LIST_ROW2) {
+              endpos = strlen(fmtchan);
+              fmtchan[endpos] = count_list[charpos];
+              fmtchan[endpos + 1] = NULL;
+              charpos++;
+            }
+            strcat(fmtchan, "\t\0");
+
+            strcat(fmtchan, chan_info[i].topic);
+            chan_ti[i] = DLLCreateNewUITextItem(fmtchan, GR_WHITE);
+            DLLOldListAddItem(chan_list, chan_ti[i]);
+            if (strcmp(oldselchan, chan_info[i].origname) == 0) {
+              DLLOldListSelectItem(chan_list, chan_ti[i]);
+            }
+          }
+          tokp = nexttok;
+          if (nexttok) {
+            nexttok = strchr(nexttok, '$');
+            if (nexttok) {
+              *nexttok = '\0';
+              nexttok++;
+            }
+          }
+          if (!tokp)
+            break;
+        }
+      }
+      DLLmem_free(pchanlist);
+    }
+
+    res = DLLPollUI();
+    // handle all UI results.
+    switch (res) {
+    case -1:
+      break;
+    case 3:
+      // Join Mastertracker game
+      {
+        DLLmprintf(0, "Sending Mastertracker game list request.\n");
+        RequestDIPGameList(DLLLocalD3Dir);
```

Comment:
Here and other places where `RequestDIPGameList()` is used. `DDLLocalD3Dir` is not always available as writable. According to logic `RequestDIPGameList` uses it as path for storing temp files. Better to rewrite it to use `std::filesystem::temp_directory_path() / Descent3` as it already used by `Descent3_temp_directory` global variable.